### PR TITLE
Review of image classes interfaces

### PR DESCRIPTION
**Pull request by heckflosse**

### CODE REVIEW

**Review by Hombre57 — APPROVED** [review]
Thanks for the `float` to `bool` fix ! Everything else is beyond my understanding, but I guess welcomed.

File: rtengine/image16.h [hunk 1]
```diff
@@ -40,62 +40,70 @@ class Image16 : public IImage16, public ImageIO
...
-    virtual void         getScanline(int row, unsigned char* buffer, int bps, bool isFloat = false);
-    virtual void         setScanline(int row, unsigned char* buffer, int bps, unsigned int numSamples);
+    virtual void getScanline(int row, unsigned char* buffer, int bps, bool isFloat = false) const;
+    virtual void setScanline(int row, unsigned char* buffer, int bps, unsigned int numSamples);
 
     // functions inherited from IImage16:
-    virtual MyMutex&     getMutex()
+    virtual MyMutex& getMutex()
     {
         return mutex();
```
Comment (Floessie) [inline rtengine/image16.h:65]: Hmm, I don't get this one (at least from the diff): Are we returning a temporary here, or is that some non-intuitive `operator ()`? And why non-const?

File: rtengine/image16.h [hunk 2]
```diff
@@ -40,62 +40,70 @@ class Image16 : public IImage16, public ImageIO
...
+    virtual int getBPS() const
     {
         return 8 * sizeof(unsigned short);
     }
-    virtual void         getScanline(int row, unsigned char* buffer, int bps, bool isFloat = false);
-    virtual void         setScanline(int row, unsigned char* buffer, int bps, unsigned int numSamples);
+    virtual void getScanline(int row, unsigned char* buffer, int bps, bool isFloat = false) const;
+    virtual void setScanline(int row, unsigned char* buffer, int bps, unsigned int numSamples);
 
     // functions inherited from IImage16:
```
Comment (Floessie) [inline rtengine/image16.h:62]: If they are inherited, better drop `virtual` and add `override`.

File: rtengine/image8.cc [hunk 3]
```diff
@@ -89,15 +89,15 @@ void Image8::setScanline (int row, unsigned char* buffer, int bps, unsigned int
     }
 }
 
-Image8* Image8::copy ()
+Image8* Image8::copy () const
```
Comment (Floessie) [inline rtengine/image8.cc:92]: Sometime in the future this should become a `std::unique_ptr<>`.

File: rtengine/imageio.cc [hunk 4]
```diff
@@ -1621,3 +1621,57 @@ int ImageIO::save (Glib::ustring fname)
...
+
+void ImageIO::getEmbeddedProfileData (int& length, unsigned char*& pdata) const
+{
+    length = loadedProfileLength;
+    pdata = (unsigned char*)loadedProfileData;
+}
+
+MyMutex& ImageIO::mutex ()
+{
+    return imutex;
```
Comment (Floessie) [inline rtengine/imageio.cc:1663]: Aha, now I get it. Crappy name. How about `MyMutex& ImageIO::getMutex() const`?

**Review by Floessie — APPROVED** [review]
:+1:, Ingo. Constification has many benefits in the long run. Could be committed as is, some optional fixes:

- De-inlining for decoupling header and implementation
- Whitespace unification (`void func()` vs. `void      func ( )`)
- `override`

Thanks for the fixes!

Best,
Flössie

File: rtengine/image16.h (hunk 2 above)
Comment (heckflosse) [reply rtengine/image16.h:62]: Done

File: rtengine/imageio.cc (hunk 4 above)
Comment (Floessie) [reply rtengine/imageio.cc:1663]: Sorry, can't be `const` when returning a non-const reference. Missed that...

Comment (heckflosse) [reply rtengine/imageio.cc:1663]: Yes, nevertheless I changed the code for part three. Will be pushed in a few minutes.

Comment (heckflosse) [reply rtengine/imageio.cc:1663]: Scratch my last comment

**Pull request merged**